abi.html

```diff
@@ -4107,73 +4106,156 @@ <h5><a href="#dso-dtor-motivation"> 3.3.6.1 Motivation </a></h5>
+The runtime library shall maintain a list of <i>global termination
+functions</i> with the following information about each:
 
 <ul>
 <li> A function pointer (a pointer to a function descriptor on Itanium).
-<li> A void* operand to be passed to the function.
-<li> A void* handle for the <i>home DSO</i> of the entry (below).
+<li> A <code>void*</code> operand to be passed to the function.
+<li> A <code>void*</code> handle for the <i>home DSO</i> of the entry (below).
 </ul>
 
 <p>
-The representation of this structure is implementation defined.
+Entries in this list need not be unique.  If the same function
+is registered multiple times, it is called multiple times as
+described below.
+
+<p>
+The runtime library shall also maintain an analogous list of
+<i>thread-local termination functions</i> for each thread, with
+the same information about each.
+
+<p>
+The representation of these lists is implementation defined.
 All references are via the API described below.
 
+<a name="dso-dtor-runtime-api"> <!-- legacy anchor -->
+<a name="global-dtor">
+<h5><a href="#global-dtor">3.3.6.3 Destructors for global objects</a></h5>
+
 <p>
-<a name="dso-dtor-runtime-api">
-<h5><a href="#dso-dtor-runtime-api"> 3.3.6.3 Runtime API </a></h5>
+After constructing an object with static storage duration
+that will require destruction on process exit, a global
+termination function is registered by calling the following
+function:
+
+<code><pre>extern "C" int __cxa_atexit(void (*f)(void *), void *p, void *d);</pre></code>
+
+<p>The third argument, <code>d</code>, must be the value
+<code>&__dso_handle</code> for the DSO which defines the object;
+see below.  The first and second arguments may be chosen at the
+implementation's convenience.
 
-<ol type=A>
 <p>
-<li> Object construction:
+The runtime library will call <code>f(p)</code> when the unloading
+of DSO <code>d</code> is required, including when the process exits.
+The call will occur before the DSO is actually unloaded from the
+process.  Calls to global termination functions will be made in the
+reverse order of their registration.
+
 <p>
-After constructing a global (or local static) object,
-that will require destruction on exit,
-a termination function is <i>registered</i> as follows:
-<center><code>
-extern "C" int __cxa_atexit ( void (*f)(void *), void *p, void *d );
-</code></center>
-This registration, e.g. <code>__cxa_atexit(f,p,d)</code>,
-is intended to cause the call <code>f(p)</code> when DSO <code>d</code> is unloaded,
-before all such termination calls registered before this one.
-It returns zero if registration is successful, nonzero on failure.
+Note that global termination functions may be registered
+concurrently.  If the registrations of two separate termination
+functions are not well-ordered by the <i>strongly happens before</i>
+relation, the order in which they are called is unspecified.
 
 <p>
-The registration function is not called from within the constructor.
```

Comment:
I removed this sentence (and an analogous sentence from the proposed thread-local section) because, as far as I can tell, they're either wrong or not really something that should be specified in the ABI:

- This certainly can't be talking about a class constructor.
- If this is talking about a global constructor function, surely the registration function *is* called from the global constructor, because I can't imagine what else would call it.
- If this is talking about a period of execution called "the constructor", well, that's not defined by the standard, and it's unclear what it would mean for the registration call to not be a part of it.
- The registration call needs to be part of the `__cxa_guard` region.  If you finish the initialization, end the guard, and then do the the registration (knowing that you just did the initialization), you lose the interthread ordering that the standard requires: the registration can race with a registration made by some other thread that has observed that the initialization is complete.